tests/assert.rs

```diff
@@ -0,0 +1,147 @@
+    }
+
+    Ok(())
+}
+
+pub async fn pod_log_contains(
+    pods: &Api<Pod>,
+    pod_name: &str,
+    expected_log: &str,
+) -> anyhow::Result<()> {
+    let logs = pods.logs(pod_name, &LogParams::default()).await?;
+    assert!(
+        logs.contains(expected_log),
+        format!("Expected log containing {} but got {}", expected_log, logs)
+    );
+    Ok(())
+}
+
+pub async fn pod_container_log_contains(
+    pods: &Api<Pod>,
+    pod_name: &str,
+    container_name: &str,
+    expected_log: &str,
+) -> anyhow::Result<()> {
+    let mut log_params = LogParams::default();
+    log_params.container = Some(container_name.to_owned());
+    let logs = pods.logs(pod_name, &log_params).await?;
+    assert!(
+        logs.contains(expected_log),
+        format!("Expected log containing {} but got {}", expected_log, logs)
+    );
+    Ok(())
+}
+
+// pub async fn pod_log_does_not_contain(
+//     pods: &Api<Pod>,
+//     pod_name: &str,
+//     unexpected_log: &str,
+// ) -> anyhow::Result<()> {
+//     let logs = pods.logs(pod_name, &LogParams::default()).await?;
+//     assert!(
+//         !logs.contains(unexpected_log),
+//         format!(
+//             "Expected log NOT containing {} but got {}",
+//             unexpected_log, logs
+//         )
+//     );
+//     Ok(())
+// }
+
+pub async fn pod_exited_successfully(pods: &Api<Pod>, pod_name: &str) -> anyhow::Result<()> {
+    let pod = pods.get(pod_name).await?;
+
+    let state = (|| {
+        pod.status?.container_statuses?[0]
+            .state
+            .as_ref()?
+            .terminated
+            .clone()
+    })()
+    .expect("Could not fetch terminated states");
+    assert_eq!(state.exit_code, 0);
+
+    Ok(())
+}
+
+pub async fn pod_exited_with_failure(pods: &Api<Pod>, pod_name: &str) -> anyhow::Result<()> {
+    let pod = pods.get(pod_name).await?;
+
+    let phase = (|| pod.status?.phase)().expect("Could not get pod phase");
+    assert_eq!(phase, "Failed");
+
+    Ok(())
+}
+
+pub async fn pod_message_contains(
+    pods: &Api<Pod>,
+    pod_name: &str,
+    expected_message: &str,
+) -> anyhow::Result<()> {
+    let pod = pods.get(pod_name).await?;
+
+    let message = (|| pod.status?.message)().expect("Could not get pod message");
+    assert!(
+        message.contains(expected_message),
+        format!(
+            "Expected pod message containing {} but got {}",
+            expected_message, message
+        )
+    );
+
+    Ok(())
+}
+
+pub async fn main_container_exited_with_failure(
+    pods: &Api<Pod>,
+    pod_name: &str,
+) -> anyhow::Result<()> {
+    let pod = pods.get(pod_name).await?;
+
+    let state = (|| {
+        pod.status?.container_statuses?[0]
+            .state
+            .as_ref()?
+            .terminated
+            .clone()
+    })()
+    .expect("Could not fetch terminated states");
+    assert_eq!(state.exit_code, 1);
+
+    Ok(())
+}
+
+pub async fn container_file_contains(
+    pod_name: &str,
```

Comment:
Super nit: You could probably make all the `&str` references in these helper functions be `AsRef<str>` for more flexibility. So:

```rust
pub async fn container_file_contains<T: AsRef<str>>(
     pod_name: T,
     pod_namespace: T,
     container_file_path: T,
     expected_content: T,
     file_error: T,
)
```

However, the `&str` here are probably fine

Reply:
After looking at this, I don't think it's worth it in this case, though it will definitely be a useful technique to have in other places - thanks!